apps/sms/test/marionette/thread_test.js

```diff
@@ -25,9 +32,74 @@ marionette('Thread Panel Tests', function() {
     );
   });
 
+  suite('General use cases', function() {
+    suite('Long SMS thread', function() {
+      var thread;
+      setup(function() {
+        thread = {
+          id: 1,
+          body: 'Thread message content',
+          lastMessageType: 'sms',
+          timestamp: Date.now(),
+          messages: [],
+          participants: ['TestMan']
+        };
+
+        for (var i = 1; i < 50; i++) {
+          thread.messages.push({
+            id: i,
+            iccId: null,
+            threadId: thread.id,
+            sender: null,
+            receiver: 'TestMan',
+            type: 'sms',
+            delivery: 'sent',
+            body: 'Thread message content ' + i,
+            timestamp: Date.now()
+          });
+        }
+
+        messagesApp.launch();
+
+        messagesApp.setStorage([thread]);
+      });
+
+      test('User can see all messages when scrolls up', function() {
+        var pageSize = 10;
+        var loadedPage = 1;
+
+        messagesApp.ThreadList.firstThread.tap();
+
+        // Composer panel should always be visible
+        assertIsDisplayed(messagesApp.Composer.sendButton);
+        assertIsDisplayed(messagesApp.Composer.messageInput);
+
+        thread.messages.forEach(function(message, index) {
+          var messageNode = messagesApp.Thread.findMessage(message.id);
+
+          // If current page is loaded, then message node should be visible
+          if (index < pageSize * loadedPage) {
```

Comment:
I think it should be good enough, we'll see if there are any intermittent over time. What I'm saying is very theorical.

when you scroll up, we display the 10 next elements that are in the DOM already; so it's theorically possible that the element we're looking for in not in the DOM yet.

But it's very theorical here, so let's forget it.